Source TikTok Marketing : Access token is null #14299

alexandr-shegeda opened the pull request:
## What
After recent changes introduced in this [PR](https://github.com/airbytehq/airbyte/pull/13917) logic related to path recalculation for JSON nodes we faced an issue for connectors that contain oneOf elements that have the same field names. So, for described case `getSortedSecretPaths()` returned duplicates, which lead to incorrect postprocessing during secret replacement.

## How
Made `getSortedSecretPaths()` return a unique values.

## Recommended reading order
1. `SecretsHelpers.java`

## 🚨 User Impact 🚨
Without these changes, users are not able to configure a proper connection using multiple auth options on spec.

## Pre-merge Checklist

<summary><strong>Updating a connector</strong></summary>

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

## Tests

<summary><strong>Unit</strong></summary>

Before fix:
<img width="1179" alt="Screenshot 2022-07-13 at 12 58 17" src="https://user-images.githubusercontent.com/10000582/178709496-8324a7c5-7d47-447c-8be1-7f6a8e5501ca.png">

<img width="1376" alt="Screenshot 2022-07-13 at 12 57 37" src="https://user-images.githubusercontent.com/10000582/178709532-0d3640d2-846f-4094-ade4-b5ec8e3d489b.png">


After fix:

<img width="773" alt="image" src="https://user-images.githubusercontent.com/10000582/178709254-bc067e9c-009d-4160-babb-a8839d7a0bd3.png">


<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

grishick commented:
because this change affects many connectors, please test that it does not break any GA and Beta connectors and also verify that it does not break jobs after a user upgrades to a platform version

sherifnada commented on airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/split_secrets/SecretsHelpers.java:
```diff
@@ -178,6 +178,7 @@ public static List<String> getSortedSecretPaths(final JsonNode spec) {
             .anyMatch(field -> field.getKey().equals(JsonSecretsProcessor.AIRBYTE_SECRET_FIELD)))
         .stream()
         .map(JsonPaths::mapJsonSchemaPathToJsonPath)
+        .distinct()
```
can you add a unit test for this?

alexandr-shegeda replied:
added unit tests. will attach screenshots with failed tests without this change

cgardens replied:
can you add in the javadoc comment that it will not return duplicates?

alexandr-shegeda commented:
> We need tests here that show the problem and then how this fixes it.

hi @cgardens, I have added unit tests for the described case, and also updated the description with failed unit tests without the fix.

cgardens reviewed (commented):
thanks! the test is great. 

can you point to what line of code causes things to work in an unexpected way when there are duplicates? this is still not intuitive to me, and i want to understand where the brittleness is coming from.

alexandr-shegeda commented:
> thanks! the test is great.
> 
> can you point to what line of code causes things to work in an unexpected way when there are duplicates? this is still not intuitive to me, and i want to understand where the brittleness is coming from.

@cgardens in case `SecretsHelpers#getSortedSecretPaths` return duplicate path
then inside of `SecretsHelpers#internalSplitAndUpdateConfig` we will replace incoming config value with the secret, and then after the second run of method JsonPaths.replaceAt()
it will not find a value of the original field, because it was replaced with {"_secret" : "..."}, so we will just erase the secret value as it showed on this screenshot <img width="1376" alt="Screenshot 2022-07-13 at 12 57 37" src="https://user-images.githubusercontent.com/10000582/178709532-0d3640d2-846f-4094-ade4-b5ec8e3d489b.png">

supertopher reviewed (approved):
Based on local testing and my understanding of the problem space this looks good to go.  

Tik Tok access token are broken so we hope to see a more different error here.

cgardens reviewed (approved):
Thank,  @alexandr-shegeda ! That explanation makes sense to me too. I think we are good to go on this.

benmoriceau commented:
> thanks! the test is great.
> 
> can you point to what line of code causes things to work in an unexpected way when there are duplicates? this is still not intuitive to me, and i want to understand where the brittleness is coming from.

@cgardens The issue is because this duplicated are secrets. Here is what happens:
- We have 2 paths for the secret that are deplicated (because they are in a oneOf)
- For each path we will try to sanitize the config in order to hide the secret and store in the secret manager
- The first one works well, store the secret in the secret store and replace the value by a coordiante
- The second on can read the secret value because it has been replace by a coordinate but doesn't error out and replace the first coordinate by a new coordinate.
- Then when we try to retrieve the secret value, it is invalid.
Does it makes sense?

supertopher reviewed (approved):
triple approve?
